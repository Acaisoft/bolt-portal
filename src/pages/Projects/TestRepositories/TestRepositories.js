import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Query, Mutation } from 'react-apollo'

import { IconButton, Typography, withStyles } from '@material-ui/core'
import { Edit, Delete } from '@material-ui/icons'

import { AddButton, DataTable, DeleteModal } from '~components'
import { forms } from '~containers'

import styles from './TestRepositories.styles'
import { GET_REPOSITORIES_QUERY } from '~services/GraphQL/Queries'
import { DELETE_REPOSITORY_MUTATION } from '~services/GraphQL/Mutations'

export class TestRepositories extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    projectId: PropTypes.string,
  }

  state = {
    open: false,
    openDeleteModal: false,
    type: null,
    updateFormValues: {
      name: null,
      url: null,
      id: null,
      projectId: this.props.projectId,
    },
    repoName: '',
    deleteRepoId: null,
  }

  handleSubmit = async (id, { delMutation }) => {
    await delMutation({ variables: { id } })
    this.setState({ openDeleteModal: false })
  }

  handleModalOpen = (id, name) => {
    this.setState({ openDeleteModal: true, repoName: name, deleteRepoId: id })
  }

  handleModalClose = () => {
    this.setState({ openDeleteModal: false })
  }

  toggleDrawer = (type, status) => {
    this.setState({
      open: status,
      type: type,
      updateFormValues: {
        name: '',
        url: '',
        projectId: this.props.projectId,
      },
    })
  }

  updateFormDrawer = (e, name, url, id) => {
    e.preventDefault()
    this.setState({
      open: true,
      type: 'update',
      updateFormValues: {
        name,
        url,
        id,
      },
    })
  }

  addTestConfigs = configs => {
    let text = ''
    configs.map((conf, i) => {
      if (configs.length === i + 1) {
        return (text += conf.name)
      } else {
        return (text += `${conf.name} | `)
      }
    })
    return text
  }

  render() {
    const { classes, projectId } = this.props
    const {
      open,
      updateFormValues,
      type,
      openDeleteModal,
      repoName,
      deleteRepoId,
    } = this.state

    return (
      <div className={classes.root}>
        <Typography variant="body2">
          Here you see results of all tests performed in project
        </Typography>
        <Mutation
          mutation={DELETE_REPOSITORY_MUTATION}
          refetchQueries={[
            { query: GET_REPOSITORIES_QUERY, variables: { projectId } },
          ]}
        >
          {(delMutation, { data }) => (
            <DeleteModal
              open={openDeleteModal}
              handleClose={this.handleModalClose}
              handleSubmit={() => this.handleSubmit(deleteRepoId, { delMutation })}
              type="repository"
              name={repoName}
            />
          )}
        </Mutation>
        <forms.Repository
          open={open}
          type={type}
          courseInitData={updateFormValues}
          close={this.toggleDrawer}
        />
        <div className={classes.btnContainer}>
          <AddButton open={this.toggleDrawer} />
        </div>
        <Query
          query={GET_REPOSITORIES_QUERY}
          variables={{ projectId }}
          fetchPolicy="cache-and-network"
        >
          {({ loading, error, data }) => {
            if (loading) return <p>Loading...</p>
            if (error) return <p>Error :(</p>
            const repositories = data.repository

            return (
              <div className={classes.tableContainer}>
                <DataTable
                  data={repositories}
                  isLoading={loading}
                  rowKey={test => test.id}
                >
                  <DataTable.Column
                    key="name"
                    render={test => test.name}
                    title="Name"
                  />
                  <DataTable.Column
                    key="url"
                    render={test => test.url}
                    title="URL"
                  />
                  <DataTable.Column
                    key="testRunType"
                    render={test => test.configurationType.name}
                    title="Test Type"
                  />
                  <DataTable.Column
                    key="testRunConfigs"
                    render={test => this.addTestConfigs(test.configurations)}
                    title="Test Run Configurations"
                  />
                  <DataTable.Column
                    key="actions"
                    render={test => (
                      <div className={classes.iconsContainer}>
                        <IconButton
                          aria-label="Edit repository"
                          className={classes.icon}
                          onClick={event =>
                            this.updateFormDrawer(
                              event,
                              test.name,
                              test.url,
                              test.id
                            )
                          }
                        >
                          <Edit />
                        </IconButton>
                        <IconButton
                          aria-label="Delete repository"
                          className={classes.icon}
                          onClick={() => this.handleModalOpen(test.id, test.name)}
                        >
                          <Delete />
                        </IconButton>
                      </div>
                    )}
                    title="Actions"
                  />
                </DataTable>
              </div>
            )
          }}
        </Query>
      </div>
    )
  }
}

export default withStyles(styles)(TestRepositories)
