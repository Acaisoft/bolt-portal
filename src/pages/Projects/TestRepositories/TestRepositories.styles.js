export default ({ palette, spacing, typography }) => {
  return {
    root: {
      flexGrow: 1,
    },
    tableContainer: {
      width: '100%',
      overflowX: 'scroll',
      marginTop: 30,
    },
    btnContainer: {
      display: 'flex',
      justifyContent: 'flex-end',
      marginBottom: 20,
    },
  }
}
