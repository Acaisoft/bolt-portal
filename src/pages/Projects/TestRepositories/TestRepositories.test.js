import React from 'react'
import { shallow } from 'enzyme'

import { TestRepositories } from './TestRepositories'
import { mocks } from '~utils/tests'

jest.mock('~containers', () => ({ forms: { Repository: 'RepositoryMock' } }))
jest.mock('~components', () => ({
  AddButton: 'AddButtonMock',
  DataTable: 'DataTableMock',
  DeleteModal: 'DeleteModalMock',
}))

const initComponent = overrides => {
  const mockProps = {
    classes: {},
  }
  const wrapper = shallow(
    <TestRepositories classes={mocks.ClassesProxy} {...mockProps} {...overrides} />
  )
  const instance = wrapper.instance()
  return { instance, wrapper }
}

describe('page: TestRepositories', () => {
  describe('rendering', () => {
    it('should render without crashing', () => {
      const { wrapper } = initComponent()
      expect(wrapper).toBeTruthy()
    })
    it('should render as expected', () => {
      const { wrapper } = initComponent()
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('events', () => {
    describe('toggleDrawer', () => {
      it('should open new-project drawer and change status state to true', () => {
        const { instance } = initComponent()
        instance.toggleDrawer('create', true)
        expect(instance.state.open).toBe(true)
        expect(instance.state.type).toBe('create')
      })
      it('should close new-project drawer and change status state to true', () => {
        const { instance } = initComponent()
        instance.toggleDrawer('create', false)
        expect(instance.state.open).toBe(false)
        expect(instance.state.type).toBe('create')
      })

      it('should close udpdate-project drawer and change status state to true', () => {
        const { instance } = initComponent()
        instance.toggleDrawer('update', false)
        expect(instance.state.open).toBe(false)
        expect(instance.state.type).toBe('update')
      })
    })
    describe('openUpdateProject', () => {
      it('should open update-project drawer with initial form data', () => {
        const { instance } = initComponent()
        const name = 'testName'
        const url = 'http://fakeUrl.com'
        const event = {
          preventDefault() {},
        }
        instance.updateFormDrawer(event, name, url)
        expect(instance.state.updateFormValues.name).toBe('testName')
        expect(instance.state.updateFormValues.url).toBe('http://fakeUrl.com')
      })
    })
    describe('addTestConfigs', () => {
      it('should create new text with divider', () => {
        const { instance } = initComponent()
        const fakeConfigs = [{ name: 'fakeName1' }, { name: 'fakeName2' }]
        const result = instance.addTestConfigs(fakeConfigs)
        expect(result).toBe('fakeName1 | fakeName2')
      })
      it('should create new text without divider', () => {
        const { instance } = initComponent()
        const fakeConfigs = [{ name: 'fakeName1' }]
        const result = instance.addTestConfigs(fakeConfigs)
        expect(result).toBe('fakeName1')
      })
    })
    describe('handleSubmit', () => {
      it('should close delete repo and close modal', () => {
        const { instance } = initComponent()
        const fakeId = 'fakeID'
        const deleteRepoMock = jest.fn()
        instance.handleSubmit(fakeId, { deleteRepoMock })
        expect(instance.state.openDeleteModal).toBe(false)
      })
    })
    describe('handleModalOpen', () => {
      it('should open modal', () => {
        const { instance } = initComponent()
        const fakeID = 'faleId'
        const fakeName = 'fakeName'
        instance.handleModalOpen(fakeID, fakeName)
        expect(instance.state.openDeleteModal).toBe(true)
      })
      it('should close modal', () => {
        const { instance } = initComponent()
        instance.handleModalClose()
        expect(instance.state.openDeleteModal).toBe(false)
      })
    })
  })
})
