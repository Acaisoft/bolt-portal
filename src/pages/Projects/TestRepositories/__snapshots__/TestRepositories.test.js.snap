// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page: TestRepositories rendering should render as expected 1`] = `
<div>
  <TypographyMock
    variant="body2"
  >
    Here you see results of all tests performed in project
  </TypographyMock>
  <Mutation
    mutation={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "deleteRepository",
            },
            "operation": "mutation",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": Object {
                    "kind": "Name",
                    "value": "repository",
                  },
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "where",
                      },
                      "value": Object {
                        "fields": Array [
                          Object {
                            "kind": "ObjectField",
                            "name": Object {
                              "kind": "Name",
                              "value": "id",
                            },
                            "value": Object {
                              "fields": Array [
                                Object {
                                  "kind": "ObjectField",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "_eq",
                                  },
                                  "value": Object {
                                    "kind": "Variable",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "id",
                                    },
                                  },
                                },
                              ],
                              "kind": "ObjectValue",
                            },
                          },
                        ],
                        "kind": "ObjectValue",
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "delete_repository",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "returning",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "uuid",
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "id",
                  },
                },
              },
            ],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 167,
          "start": 0,
        },
      }
    }
    refetchQueries={
      Array [
        Object {
          "query": Object {
            "definitions": Array [
              Object {
                "directives": Array [],
                "kind": "OperationDefinition",
                "name": Object {
                  "kind": "Name",
                  "value": "getRepositories",
                },
                "operation": "query",
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "where",
                          },
                          "value": Object {
                            "fields": Array [
                              Object {
                                "kind": "ObjectField",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "project_id",
                                },
                                "value": Object {
                                  "fields": Array [
                                    Object {
                                      "kind": "ObjectField",
                                      "name": Object {
                                        "kind": "Name",
                                        "value": "_eq",
                                      },
                                      "value": Object {
                                        "kind": "Variable",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "projectId",
                                        },
                                      },
                                    },
                                  ],
                                  "kind": "ObjectValue",
                                },
                              },
                            ],
                            "kind": "ObjectValue",
                          },
                        },
                      ],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "repository",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "id",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "name",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "url",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "configurations",
                            },
                            "selectionSet": Object {
                              "kind": "SelectionSet",
                              "selections": Array [
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "id",
                                  },
                                  "selectionSet": undefined,
                                },
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "name",
                                  },
                                  "selectionSet": undefined,
                                },
                              ],
                            },
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "configurationType",
                            },
                            "selectionSet": Object {
                              "kind": "SelectionSet",
                              "selections": Array [
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "id",
                                  },
                                  "selectionSet": undefined,
                                },
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "name",
                                  },
                                  "selectionSet": undefined,
                                },
                              ],
                            },
                          },
                        ],
                      },
                    },
                  ],
                },
                "variableDefinitions": Array [
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NamedType",
                      "name": Object {
                        "kind": "Name",
                        "value": "uuid",
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "projectId",
                      },
                    },
                  },
                ],
              },
            ],
            "kind": "Document",
            "loc": Object {
              "end": 259,
              "start": 0,
            },
          },
          "variables": Object {
            "projectId": undefined,
          },
        },
      ]
    }
  >
    <Component />
  </Mutation>
  <RepositoryFormMock
    close={[Function]}
    courseInitData={
      Object {
        "id": null,
        "name": null,
        "projectId": undefined,
        "url": null,
      }
    }
    open={false}
    type={null}
  />
  <div>
    <AddButtonMock
      open={[Function]}
    />
  </div>
  <Query
    fetchPolicy="cache-and-network"
    query={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "getRepositories",
            },
            "operation": "query",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "where",
                      },
                      "value": Object {
                        "fields": Array [
                          Object {
                            "kind": "ObjectField",
                            "name": Object {
                              "kind": "Name",
                              "value": "project_id",
                            },
                            "value": Object {
                              "fields": Array [
                                Object {
                                  "kind": "ObjectField",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "_eq",
                                  },
                                  "value": Object {
                                    "kind": "Variable",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "projectId",
                                    },
                                  },
                                },
                              ],
                              "kind": "ObjectValue",
                            },
                          },
                        ],
                        "kind": "ObjectValue",
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "repository",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "name",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "url",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "configurations",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "configurationType",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "uuid",
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "projectId",
                  },
                },
              },
            ],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 259,
          "start": 0,
        },
      }
    }
    variables={
      Object {
        "projectId": undefined,
      }
    }
  >
    <Component />
  </Query>
</div>
`;
